panel/chat/feed.py

```diff
@@ -496,7 +528,10 @@ async def _handle_callback(self, message, loop: asyncio.BaseEventLoop):
             response = self.callback(*callback_args)
         else:
             self._callback_future = None
-            response = await asyncio.to_thread(self.callback, *callback_args)
+            if isgeneratorfunction(self.callback):
+                response = self._to_async_gen(self.callback(*callback_args))
```

Comment:
This doesn't look quite right as is, doesn't there have to be an await somewhere? I'm actually a little confused how generators work in here. How do the yielded values become responses?

Reply:
Actually, I just tested; there's no need to introduce complexity because serialize response handles the generator fine.

This test should also catch failures.
https://github.com/holoviz/panel/pull/6312/files#diff-071d65f945462dbf3517ddd97071c3da199d35c6f6a74174dbf3eec329826808R530-R531

<img width="1307" alt="image" src="https://github.com/holoviz/panel/assets/15331990/9aaae109-c8d6-41ca-9cd6-a378de78f250">


```python
    async def _serialize_response(self, response: Any) -> ChatMessage | None:
        """
        Serializes the response by iterating over it and
        updating the message's value.
        """
        response_message = None
        try:
            if isasyncgen(response):
                self._callback_state = CallbackState.GENERATING
                async for token in response:
                    response_message = self._upsert_message(token, response_message)
                    response_message.show_activity_dot = self.show_activity_dot
            elif isgenerator(response):
                self._callback_state = CallbackState.GENERATING
                for token in response:
                    response_message = self._upsert_message(token, response_message)
                    response_message.show_activity_dot = self.show_activity_dot
            elif isawaitable(response):
                response_message = self._upsert_message(await response, response_message)
            else:
                response_message = self._upsert_message(response, response_message)
        finally:
            if response_message:
                response_message.show_activity_dot = False
        return response_message

```

Reply:
It's still blocking though right?

Reply:
Yes it's still blocking. Based on that, I think I should just be calling it as a normal function...? https://github.com/holoviz/panel/pull/6312/commits/7f5af91ca002863f5c83ed30580a08645688176b